### PR TITLE
feature: Check lock in TC when use @GlobalLock

### LINKED ISSUE #3067: Check lock in TC when use @GlobalLock
## Why you need it?
For now if you want to use @GlobalLock, you must use it with @Transactional, and check the global lock when transaction commit.
But it will cause two problems:
1. Can not check global lock when use @GlobalLock Independently.
2. Check global lock can not retry.

If you want to fix it, please comment "Please assign it to me" is this issue.
## Other related information

### PR DESCRIPTION
**Pull request by caohdgege**
2. Check global lock retry when conflict.

### Ⅰ. Describe what this PR did
1. make @GlobalLock work Independently
2. Check global lock can retry

### Ⅱ. Does this pull request fix one issue?

fixes #3067 

### Ⅳ. Describe how to verify it
can check lock when only use @GLobalLock。
can retry check lock when use @GlobalLock。

### CODE REVIEW

**Comment by a364176773** [comment]
[ERROR] /home/travis/build/seata/seata/rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java:28:8: Unused import - io.seata.rm.datasource.ConnectionProxy. [UnusedImports]
[ERROR] /home/travis/build/seata/seata/rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java:22:8: Unused import - java.util.concurrent.Callable. [UnusedImports]
[ERROR] /home/travis/build/seata/seata/rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java:26:8: Unused import - io.seata.rm.datasource.ConnectionContext. [UnusedImports]
[ERROR] /home/travis/build/seata/seata/rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:29: Using the '.*' form of import should be avoided - io.seata.rm.datasource.*. [AvoidStarImport]

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java [hunk 1]
```diff
@@ -113,7 +111,13 @@ public T doExecute(Object... args) throws Throwable {
                     } else {
                         conn.rollback();
                     }
-                    lockRetryController.sleep(lce);
+
+                    // retry hear only when the config is true
+                    if (LockRetryPolicy.isLockRetryPolicyBranchRollbackOnConflict()) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java:116]: it's not consistent with the previous meaning

Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java]: change

Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java]: change

**Comment by ggndnn** [comment]
@caohdgege 
Similar ideas as #1649?

**Comment by caohdgege** [comment]
> @caohdgege
> Similar ideas as #1649?

帮忙把 #1649 关联一下#3067，然后把这个pr关掉吧

**Review by slievrly — APPROVED** [review]
LGTM

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java [hunk 2]
```diff
@@ -96,13 +96,11 @@ public T doExecute(Object... args) throws Throwable {
                         break;
                     }
 
-                    if (RootContext.inGlobalTransaction()) {
-                        //do as usual
+                    if (RootContext.inGlobalTransaction() || RootContext.requireGlobalLock()) {
+                        // do as usual
+                        // in @GlobalLock env, don't need to append key eithor,
+                        // but check lock hear, do the same thing as @GlobalTransactional env
```
Comment (ggndnn) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/SelectForUpdateExecutor.java:102]: Is it better to comment the code like **`"Do the same thing under either @GlobalTransactional or @GlobalLock, that only check the global lock here."`** or something else more accurate? There is nothing about **`“key appending”`** in the new code block, so the comment could be a little puzzling.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java [hunk 3]
```diff
@@ -407,5 +407,4 @@ protected TableRecords buildTableRecords(Map<String, List<Object>> pkValuesMap)
     protected String getDbType() {
         return statementProxy.getConnectionProxy().getDbType();
     }
-
```
Comment (ggndnn) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:410]: What's this line deleting for? Is it necessary?

Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:410]: 空白行，应该是之前做代码变更的时候不小心删的

Comment (ggndnn) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:410]: Can you revert this modification?

Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java]: ok

**Comment by ggndnn** [comment]
@caohdgege same question as https://github.com/seata/seata/pull/1649#discussion_r324210332 so far, `@globalCheck` could only prevent "select for update in current global tx" from reading data locked by  "another global tx" but could not prevent "another global tx" from visting data locked by  "select for update in current global tx" all the time, because db row lock will be released after local tx is committed. Do you think that should be handled or ignored in this pr?

**Comment by caohdgege** [comment]
> @caohdgege same question as [#1649 (comment)](https://github.com/seata/seata/pull/1649#discussion_r324210332) so far, `@globalCheck` could only prevent "select for update in current global tx" from reading data locked by "another global tx" but could not prevent "another global tx" from visting data locked by "select for update in current global tx" all the time, because db row lock will be released after local tx is committed. Do you think that should be handled or ignored in this pr?

我刚开始的实现上是想，如果是在GlobalTransactional的情况下，selectForUpdate也去acquire lock的，这样的话就可以解决你说的这个问题。但是后面跟 @l81893521 嘉伟哥讨论的时候，说这样可能到拖慢性能，可能需要社区讨论一下，本次pr不涉及解决这个问题。
然后我之前也提了个issue是关于GlobalTransactional的情况下应该要去acquire lock的，我们可以在那边讨论这个问题。#3075

**Comment by ggndnn** [comment]
> > @caohdgege same question as [#1649 (comment)](https://github.com/seata/seata/pull/1649#discussion_r324210332) so far, `@globalCheck` could only prevent "select for update in current global tx" from reading data locked by "another global tx" but could not prevent "another global tx" from visting data locked by "select for update in current global tx" all the time, because db row lock will be released after local tx is committed. Do you think that should be handled or ignored in this pr?
> 
> 我刚开始的实现上是想，如果是在GlobalTransactional的情况下，selectForUpdate也去acquire lock的，这样的话就可以解决你说的这个问题。但是后面跟 @l81893521 嘉伟哥讨论的时候，说这样可能到拖慢性能，可能需要社区讨论一下，本次pr不涉及解决这个问题。
> 然后我之前也提了个issue是关于GlobalTransactional的情况下应该要去acquire lock的，我们可以在那边讨论这个问题。#3075

Fine.

**Review by ggndnn — APPROVED** [review]
LGTM.

**Pull request merged**